File: app/schemas/system.py
from pydantic import BaseModel, EmailStr, Field, field_validator


class AdminUpdateStatus(BaseModel):
    enabled: bool
    running: bool
    ref: str
    log_tail: list[str]


class TelegramSettingsRead(BaseModel):
    bot_username: str | None
    bot_token_set: bool
    admin_telegram_id: str | None
    socks5_enabled: bool
    socks5_host: str | None
    socks5_port: int | None
    socks5_username_set: bool
    admin_email: EmailStr


class TelegramTestMessageResult(BaseModel):
    ok: bool
    message: str


class TelegramSettingsUpdate(BaseModel):
    bot_username: str | None = Field(default=None, max_length=64)
    bot_token: str | None = Field(default=None, max_length=256)
    admin_telegram_id: str | None = Field(default=None, max_length=32)
    socks5_enabled: bool = False
    socks5_host: str | None = Field(default=None, max_length=255)
    socks5_port: int | None = Field(default=None, ge=1, le=65535)
    socks5_username: str | None = Field(default=None, max_length=255)
    socks5_password: str | None = Field(default=None, max_length=255)
    admin_email: EmailStr | None = None
    current_password: str | None = Field(default=None, min_length=8, max_length=256)
    new_password: str | None = Field(default=None, min_length=8, max_length=256)

    @field_validator(
        "bot_username",
        "bot_token",
        "admin_telegram_id",
        "socks5_host",
        "socks5_username",
        "socks5_password",
        "current_password",
        "new_password",
        mode="before",
    )
    @classmethod
    def empty_string_to_none(cls, value: str | None) -> str | None:
        if isinstance(value, str):
            stripped = value.strip()
            return stripped or None
        return value

File: frontend/src/api/system.ts
import { api } from "./client";

export type AdminUpdateStatus = {
  enabled: boolean;
  running: boolean;
  ref: string;
  log_tail: string[];
};

export type DiagnosticCheckRead = {
  name: string;
  ok: boolean;
  message: string;
  fix: string | null;
};

export type DiagnosticLogRead = {
  service: string;
  lines: string[];
  error: string | null;
};

export type DiagnosticsRead = {
  checks: DiagnosticCheckRead[];
  logs: DiagnosticLogRead[];
};

export type TelegramSettingsRead = {
  bot_username: string | null;
  bot_token_set: boolean;
  admin_telegram_id: string | null;
  socks5_enabled: boolean;
  socks5_host: string | null;
  socks5_port: number | null;
  socks5_username_set: boolean;
  admin_email: string;
};

export type TelegramSettingsUpdate = {
  bot_username?: string;
  bot_token?: string;
  admin_telegram_id?: string;
  socks5_enabled: boolean;
  socks5_host?: string;
  socks5_port?: number;
  socks5_username?: string;
  socks5_password?: string;
  admin_email?: string;
  current_password?: string;
  new_password?: string;
};

export type TelegramTestMessageResult = {
  ok: boolean;
  message: string;
};

export async function getUpdateStatus(): Promise<AdminUpdateStatus> {
  const response = await api.get<AdminUpdateStatus>("/system/updates");
  return response.data;
}

export async function startUpdate(): Promise<AdminUpdateStatus> {
  const response = await api.post<AdminUpdateStatus>("/system/updates");
  return response.data;
}

export async function getDiagnostics(params?: {
  services?: string[];
  tail?: number;
}): Promise<DiagnosticsRead> {
  const response = await api.get<DiagnosticsRead>("/system/diagnostics", {
    params: {
      services: params?.services,
      tail: params?.tail,
    },
    paramsSerializer: { indexes: null },
  });
  return response.data;
}

export async function getTelegramSettings(): Promise<TelegramSettingsRead> {
  const response = await api.get<TelegramSettingsRead>("/system/telegram-settings");
  return response.data;
}

export async function updateTelegramSettings(
  payload: TelegramSettingsUpdate,
): Promise<TelegramSettingsRead> {
  const response = await api.put<TelegramSettingsRead>("/system/telegram-settings", payload);
  return response.data;
}

export async function sendTelegramTestMessage(): Promise<TelegramTestMessageResult> {
  const response = await api.post<TelegramTestMessageResult>(
    "/system/telegram-settings/test-message",
  );
  return response.data;
}
